src/librustc_errors/emitter.rs

```diff
@@ -893,7 +893,8 @@ impl EmitterWriter {
         } else {
             buffer.append(0, &level.to_string(), Style::Level(level.clone()));
             match code {
-                &Some(ref code) => {
+                // only render error codes, not lint codes
+                &Some(ref code) if code.starts_with("E") && code.len() == 5 => {
```

Comment:
Might be better to check for the regex `E[0-9]+` - I could imagine someone using a string "Exlint" and getting confused about the weird rendering.

Reply:
You can't do that. they are lowercase by definition (autogenerated from the global-name in the lint declaration macro)

Reply:
@oli-obk 

> You can't do that. they are lowercase by definition (autogenerated from the global-name in the lint declaration macro)

Wait, I'm confused -- what are you saying here? The error codes are lowercase? You're already checking for `E`, so that can't be it.

Still, this kinda' feels like a big hack to me. Why not add another field to the diagnostic, like `lint_name` or something? It seems pretty odd that our final JSON winds up with `"explanation":""`, which has nothing to do with lints.

Reply:
I meant lint names are always lowercase, that's how we can distinguish them.

That said, I can replace the `code` field (currently a `String`) with an

```rust
enum DiagnosticId {
    Error(String),
    Lint(String),
}
```

> It seems pretty odd that our final JSON winds up with "explanation":"", which has nothing to do with lints.

The explanation being empty is an implementation detail. It was nontrivial to add and I didn't want to mess up the PR with unrelated changes. The issue is that the array with the error explanations is filled in *before* the lints are loaded into the session and then frozen by creating the error emitter. So once the lints are loaded (and thus their explanations are available), we can't place them into the emitter.

It just requires some refactoring to get done.

I found it odder to have two `Option` fields which are mutually exclusive than a temporarily empty explanation string.